### ISSUE TITLE
Feature request: Covariance and Contravariance support for interfaced properties

### ISSUE THREAD

**Issue by Wenzil** [jacobdufault/fullinspector#9]
This is more feature request than an issue, feel free to close this if this is not the appropriate place for it.

The topic of covariance and contravariance is quite counter-intuitive so hopefully my example will make some sense. The "in" is really important. Say I have the following definitions.

``` csharp
[JsonObject(MemberSerialization.OptIn)]
public interface IAction<in T>
{
    void Act(T data);
}

[JsonObject(MemberSerialization.OptIn)]
public class DestroyGameObject : IAction<object>
{
    [JsonProperty]
    private GameObject gameObject;

    public void Act(object data)
    {
        // we ignore the data object in this case, just destroy the game object
        // but you can imagine a PrintMessage : IAction<string> action which prints
        // the given (at runtime) message
        GameObject.Destroy(gameObject);
    }
}
```

Given this, because T is contravariant in IAction<T>, the following assignment is valid: (and that makes sense because DestroyGameObject.Act() should be able to accept a string parameter if it expects an object)

``` csharp
[JsonProperty]
IAction<string> action = new DestroyGameObject();
```

In other words, DestroyGameObject is a valid implementation of IAction<string>, and so should appear in the dropdown list of concrete implementations of a IAction<string> property in the inspector. 

I imagine that making this work would only be a matter of doing more precise reflection by considering covariance and contravariance. More info on covariance and contravariance here:
http://msdn.microsoft.com/en-us/library/dd799517(v=vs.110).aspx
